lib/agent0/agent0/hyperdrive/policies/lpandarb.py

```diff
@@ -0,0 +1,642 @@
+"""Agent policy for LP trading that also arbitrage on the fixed rate."""
+from __future__ import annotations
+
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from fixedpointmath import FixedPoint
+
+from agent0.hyperdrive.state import HyperdriveActionType, HyperdriveMarketAction
+from elfpy.types import MarketType, Trade
+
+from .hyperdrive_policy import HyperdrivePolicy
+
+if TYPE_CHECKING:
+    from ethpy.hyperdrive import HyperdriveInterface
+    from numpy.random._generator import Generator as NumpyGenerator
+
+    from agent0.hyperdrive.state import HyperdriveWallet
+
+import time
+
+# pylint: disable=too-many-parameters, too-many-local-variables
+
+# constants
+fp0 = FixedPoint(0)
+fp1 = FixedPoint(1)
+fp2 = FixedPoint(2)
+fp12 = FixedPoint(12)
+fp_seconds_in_year = FixedPoint(365 * 24 * 60 * 60)
+tolerance = 1e-18
+MAX_ITER = 10
+MIN_TRADE_AMOUNT = FixedPoint(scaled_value=20)
+
+
+# functions
+def calc_bond_reserves(
+    share_reserves: FixedPoint,
+    share_price: FixedPoint,
+    target_rate: FixedPoint,
+    position_duration: FixedPoint,
+    time_stretch: FixedPoint,
+):
+    """Calculate the amount of bonds that hit the target target rate for the given shares.
+
+    Parameters
+    ----------
+    share_reserves : FixedPoint
+        The amount of share reserves.
+    share_price : FixedPoint
+        The price of the share.
+    target_rate : FixedPoint
+        The target rate.
+    position_duration : FixedPoint
+        The duration of the position.
+    time_stretch : FixedPoint
+        The time stretch factor.
+
+    Returns
+    -------
+    FixedPoint
+        The amount of bonds that hit the target rate.
+    """
+    return share_price * share_reserves * ((fp1 + target_rate * position_duration / fp_seconds_in_year) ** time_stretch)
+
+
+def calc_spot_price_local(
```

Comment:
I assume you're doing this because you don't want to use the actual pool config & pool info on-chain (as in you want to get spot price given some trade mock-up).

FWIW, pyperdrive has `pyperdrive.get_spot_price(pool_config_str, pool_info_str)`. It's not ideal bc you have to construct the pool info & config objects. But also we really want to move these calcs over to the rust sdk.

For now you could do one of two things:
- I think it's fine to just add a TODO to use the pyperdrive functions when we clean this up.
- or if you want to do it better now, you could access the current `interface.pool_info` and `interface.pool_config`, modify them as you make mock trades, serialize them (e.g. by making `_serialized_pool_config` and `_serialized_pool_info` public), and pass them to the `pyperdrive.get_spot_price()` function.

Reply:
these are used in the internal functions where we don't currently have pool_config and pool_info, so i'm going to just put a TODO for now